In [45]:
import pandas as pd
import numpy as np

# Read a comma-separated values (csv) file into DataFrame.

The data folder contains a file named "newark_tree_data.csv". Read this file into a dataframe.


resources:
* [Pandas read csv docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas.read_csv)

In [8]:
# Create DataFrame here.
df = pd.read_csv('data/newark_tree_data.csv')

In [9]:
# Display the first few rows of the DataFrame here.
df.head()

ID         ON       AREA  ADDRESS       STREET  SITE   SIDE  ASSIGNED  \
0  11.0  Elan Park  Elan Park        2  Blue Hen Rg  11.0  Front     False   
1  12.0  Elan Park  Elan Park        2  Blue Hen Rg  12.0  Front     False   
2  13.0  Elan Park  Elan Park        2  Blue Hen Rg  13.0  Front     False   
3  14.0  Elan Park  Elan Park        2  Blue Hen Rg  14.0  Front     False   
4  15.0  Elan Park  Elan Park        2  Blue Hen Rg  15.0  Front     False   

                 COMMON               BOTANICAL  ... IMPROPMULC  IMPROPPRUN  \
0   arborvitae, eastern      Thuja occidentalis  ...      False       False   
1   arborvitae, eastern      Thuja occidentalis  ...      False       False   
2   pine, eastern white           Pinus strobus  ...      False       False   
3  whitecedar, Atlantic  Chamaecyparis thyoides  ...      False       False   
4  whitecedar, Atlantic  Chamaecyparis thyoides  ...      False       False   

  NUTRIENTD   PEST GIRDLING STRESS  OHUTILITY  TOP_HEIGHT CROWN_M CROWN_DIE  
0     False  False    False  False         No          40    0-5%      0-5%  
1     False  False    False  False         No          41    0-5%      0-5%  
2     False  False     True  False         No          54    0-5%     6-10%  
3     False  False     True  False         No          57  11-15%    26-30%  
4     False  False     True  False         No          55  11-15%    21-25%  

[5 rows x 54 columns]

In [32]:
# How many columns does the DataFrame contain?
df.shape

(4193, 54)

In [29]:
# What attributes (columns) does this DataFrame contain?
df.columns

Index(['ID', 'ON', 'AREA', 'ADDRESS', 'STREET', 'SITE', 'SIDE', 'ASSIGNED',
       'COMMON', 'BOTANICAL', 'CULTIVAR', 'DBH', 'REMOVAL', 'STEMS', 'COND',
       'MAINTPRIM', 'OBSERVE', 'INSPECT', 'LOCATION', 'NOTES', 'DATE', 'TIME',
       'STAFF', 'XCOORD', 'YCOORD', 'FAILURE', 'LIKELIHOOD', 'TARGET',
       'CONSEQ', 'RISK', 'UNIQUEID', 'EDITTIME', 'SPREAD', 'HEIGHT',
       'CLEARANCE', 'DECAYCL', 'DECAYT', 'DECAYRF', 'DAMC', 'DAMT', 'DAMRF',
       'CAVITYCL', 'CAVITYT', 'CAVITYRF', 'IMPROPMULC', 'IMPROPPRUN',
       'NUTRIENTD', 'PEST', 'GIRDLING', 'STRESS', 'OHUTILITY', 'TOP_HEIGHT',
       'CROWN_M', 'CROWN_DIE'],
      dtype='object')

Upon reviewing the columns of the DataFrame, we see that there is already a column named ID that is well suited for an index column.  

For the purposes of this lab, we are only going to need a few of the available attributes. The attributes we wish to keep are listed below:
* ID
* UNIQUEID
* COMMON
* BOTANICAL
* COND
* RISK
* HEIGHT

Let's provide the index_col and the usecols args when we invoke the read_csv method.

In [63]:
# Create the new DataFrame here.
df2 = pd.read_csv('data/newark_tree_data.csv', usecols=["ID", "UNIQUEID", "COMMON", "BOTANICAL", "COND", "RISK", "HEIGHT"])
df2 = df2.set_index("ID")

In [64]:
# Display the first few rows of the new DataFrame here.
df2.head()

COMMON               BOTANICAL       COND RISK  \
ID                                                                   
11.0   arborvitae, eastern      Thuja occidentalis       Good  Low   
12.0   arborvitae, eastern      Thuja occidentalis  Very Good  Low   
13.0   pine, eastern white           Pinus strobus       Good  Low   
14.0  whitecedar, Atlantic  Chamaecyparis thyoides       Fair  Low   
15.0  whitecedar, Atlantic  Chamaecyparis thyoides       Fair  Low   

               UNIQUEID  HEIGHT  
ID                               
11.0  LM 20150929141218      40  
12.0  LM 20150929141418      40  
13.0  LM 20150929141844      54  
14.0  LM 20150929142424      56  
15.0  LM 20150929143018      55

The data narritive specifies that 4,193 sites were inventoried. Use the DataFrame shape attribute to verify the number of rows.

In [69]:
# How many rows does your DataFrame contain.
df2.shape

(4193, 6)

Let's look at the distribution of data for the HEIGHT attribute.

In [70]:
# What is the min height?
df2["HEIGHT"].min()

0

In [71]:
# What is the max height?
df2["HEIGHT"].max()

140

In [73]:
# How do we acquire count, mean, standard deviation, and distribution information from the DataFrame 
# by invoking a single method?
df2.describe()

HEIGHT
count  4193.000000
mean     55.311472
std      36.293489
min       0.000000
25%      23.000000
50%      50.000000
75%      86.000000
max     140.000000

# Plot a histogram using the HEIGHT attribute. 

resources:
* [Magic commands](https://ipython.readthedocs.io/en/stable/interactive/magics.html)
* [Pandas Series Hist](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.plot.hist.html#pandas.Series.plot.hist)

In [ ]:
# Apply magic command to display matplotlib images inline


In [ ]:
# Create a histogram using the HEIGHT attribute here


The data narrative specifies the following:
"A total of 4,193 sites were inventoried, including 4,087 trees, 100 stumps, and 6 vacant planting sites"

Earlier, we were able to validate the total number of entries in the DataFrame. Let's now dig in a bit deeper and validate the stumps and vacant planting sites.

In [ ]:
# How many entries are stumps


In [ ]:
# How many entries are vacant sites


In [ ]:
# Do these number add up?


# Plot a bar graph using the COND attribute.

resources:
* [Pandas Series Bar](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.plot.bar.html#pandas.Series.plot.bar)

In [ ]:
# How can we determine all of the possible values describing the condition of a tree and their associated counts?


In [ ]:
# Were these already in the order you would want them to be in order for a graph to make sense? 
# If not, reorder  the Series.


In [ ]:
# Plot the Series here


In [ ]:
# How many entries have missing values for the COND attribute? Why are these values missing?


# Plot a pie chart using the RISK attribute.

resources:
* [Pandas Series Pie](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.plot.pie.html#pandas.Series.plot.pie)
* [matplotlib pie chart docs](https://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.pie)

In [ ]:
# How can we determine all of the possible values describing the RISK associated with each tree
# along with the corresponding row counts?


In [ ]:
# Create an array containing the colors green, yellow, and orange
# When we create our pie chart, the colors should be associated with the attributes according to the specifications below:
# Green -> Low
# Yellow -> Moderate
# Orange -> High


In [ ]:
# Plot the pie chart here


In [ ]:
# How many entries have missing values for the RISK attribute? Why are these values missing?


# Indexing

In [ ]:
# Display row by position 1


In [ ]:
# Display row by label 1
